examples/keras_io/tensorflow/vision/knowledge_distillation.py

```diff
@@ -0,0 +1,293 @@
+effectively smoothing out the probability distribution and revealing
+inter-class relationships learned by the teacher.
+
+**Reference:**
+
+- [Hinton et al. (2015)](https://arxiv.org/abs/1503.02531)
+"""
+
+"""
+## Setup
+"""
+
+import os
+
+os.environ["KERAS_BACKEND"] = "tensorflow"
+
+import keras
+from keras import layers
+from keras import ops
+import tensorflow as tf
+import numpy as np
+
+"""
+## Construct `Distiller()` class
+
+The custom `Distiller()` class, overrides the `Model` methods `train_step`, `test_step`,
+and `compile()`. In order to use the distiller, we need:
+
+- A trained teacher model
+- A student model to train
+- A student loss function on the difference between student predictions and ground-truth
+- A distillation loss function, along with a `temperature`, on the difference between the
+soft student predictions and the soft teacher labels
+- An `alpha` factor to weight the student and distillation loss
+- An optimizer for the student and (optional) metrics to evaluate performance
+
+In the `train_step` method, we perform a forward pass of both the teacher and student,
+calculate the loss with weighting of the `student_loss` and `distillation_loss` by
+`alpha` and
+`1 - alpha`, respectively, and perform the backward pass. Note: only the student weights
+are updated,
+and therefore we only calculate the gradients for the student weights.
+
+In the `test_step` method, we evaluate the student model on the provided dataset.
+"""
+
+
+class Distiller(keras.Model):
+    def __init__(self, student, teacher):
+        super().__init__()
+        self.teacher = teacher
+        self.student = student
+
+    def compile(
+        self,
+        optimizer,
+        metrics,
+        student_loss_fn,
+        distillation_loss_fn,
+        alpha=0.1,
+        temperature=3,
+    ):
+        """Configure the distiller.
+
+        Args:
+            optimizer: Keras optimizer for the student weights
+            metrics: Keras metrics for evaluation
+            student_loss_fn: Loss function of difference between student
+                predictions and ground-truth
+            distillation_loss_fn: Loss function of difference between soft
+                student predictions and soft teacher predictions
+            alpha: weight to student_loss_fn and 1-alpha to distillation_loss_fn
+            temperature: Temperature for softening probability distributions.
+                Larger temperature gives softer distributions.
+        """
+        super().compile(optimizer=optimizer, metrics=metrics)
+        self.student_loss_fn = student_loss_fn
+        self.distillation_loss_fn = distillation_loss_fn
+        self.alpha = alpha
+        self.temperature = temperature
+
+    def train_step(self, data):
+        # Unpack data
+        x, y = data
+
+        # Forward pass of teacher
+        teacher_predictions = self.teacher(x, training=False)
+
+        with tf.GradientTape() as tape:
```

Comment:
This is the only use of `tf` in the example. That means it's pretty close to being able to run on any framework.

The only train_step customization here is the computation of a custom loss. I believe it may be implemented in a fully backend-agnostic way by overriding `.compute_loss()`, which is a method shared across all backends.

The forward pass (`call()`) can be redirected to `student.call`, and `student_predictions` will be computed outside of `compute_loss()`. The signature is `compute_loss(self, x=None, y=None, y_pred=None, **kwargs)` -- x is the input, y_pred will be student_predictions.

Reply:
Thanks for tip. I've made the adjustments and now it is backend agnostic . I tried with `jax` and ran smoothly. 

Btw noticed a weird thing, `compute_loss(self, x=None, y=None, y_pred=None, **kwargs)` throws error with `jax` saying `number of arguments mismatch (takes 4 but 5 given)`. However, setting `compute_loss(self, x=None, y=None, y_pred=None, sample_weight=None, allow_empty=False)` solved the issue

Reply:
Also as code is backend agnostic now, should I move the code to `./vision` from `./tensorflow/vision`?

Reply:
I guess compiled JAX functions don't handle `kwargs` well. Better to keep the full signature then.

Did you try with torch as well?

Yes, the example can be moved out of the `tensorflow` folder now.

Reply:
I'll move it to `./vision` then. Meanwhile, I'll also check if it works with `torch`

Reply:
@fchollet just checked, works fine with `torch` even with `**kwargs`.

Reply:
I have also moved the code to `/vision`